iothub_client/src/iothubtransport_mqtt_common.c

```diff
@@ -4006,32 +4025,57 @@ STRING_HANDLE IoTHubTransport_MQTT_Common_GetHostname(TRANSPORT_LL_HANDLE handle
     return result;
 }
 
-IOTHUB_CLIENT_RESULT IoTHubTransport_MQTT_Common_SendMessageDisposition(MESSAGE_CALLBACK_INFO* message_data, IOTHUBMESSAGE_DISPOSITION_RESULT disposition)
+IOTHUB_CLIENT_RESULT IoTHubTransport_MQTT_Common_SendMessageDisposition(IOTHUB_DEVICE_HANDLE handle, IOTHUB_MESSAGE_HANDLE messageHandle, IOTHUBMESSAGE_DISPOSITION_RESULT disposition)
 {
-    (void)disposition;
-
     IOTHUB_CLIENT_RESULT result;
-    if (message_data)
+    if (handle == NULL || messageHandle == NULL)
+    {
+        /*Codes_SRS_IOTHUB_MQTT_TRANSPORT_10_001: [If messageData is NULL, IoTHubTransport_MQTT_Common_SendMessageDisposition shall fail and return IOTHUB_CLIENT_ERROR. ]*/
+        LogError("Invalid argument (handle=%p, messageHandle=%p", handle, messageHandle);
+        result = IOTHUB_CLIENT_INVALID_ARG;
+    }
+    else if (disposition == IOTHUBMESSAGE_ASYNC_ACK)
+    {
+        LogError("IOTHUBMESSAGE_ASYNC_ACK is not a valid disposition value for this function");
+        result = IOTHUB_CLIENT_INVALID_ARG;
+    }
+    else
     {
-        if (message_data->messageHandle)
+        if (disposition == IOTHUBMESSAGE_ACCEPTED)
         {
-            IoTHubMessage_Destroy(message_data->messageHandle);
-            result = IOTHUB_CLIENT_OK;
+            MESSAGE_DISPOSITION_CONTEXT* msgDispCtx = NULL;
+
+            if (IoTHubMessage_GetDispositionContext(messageHandle, &msgDispCtx) != IOTHUB_MESSAGE_OK ||
+                msgDispCtx == NULL)
+            {
+                /* Codes_SRS_TRANSPORTMULTITHTTP_10_002: [If any of the messageData fields are NULL, IoTHubTransportHttp_SendMessageDisposition shall fail and return IOTHUB_CLIENT_ERROR.] */
+                LogError("invalid message handle (no disposition context found)");
+                result = IOTHUB_CLIENT_ERROR;
+            }
+            else
+            {
+                PMQTTTRANSPORT_HANDLE_DATA transport_data = (PMQTTTRANSPORT_HANDLE_DATA)handle;
+
+                if (mqtt_client_send_message_response(transport_data->mqttClient, msgDispCtx->packet_id, msgDispCtx->qos_value) != 0)
+                {
+                    LogError("Failed sending ACK for MQTT message (packet_id=%u)", msgDispCtx->packet_id);
+                    result = IOTHUB_CLIENT_ERROR;
+                }
+                else
+                {
+                    result = IOTHUB_CLIENT_OK;
+                }
+            }
         }
         else
         {
-            /*Codes_SRS_IOTHUB_MQTT_TRANSPORT_10_002: [If any of the messageData fields are NULL, IoTHubTransport_MQTT_Common_SendMessageDisposition shall fail and return IOTHUB_CLIENT_ERROR. ]*/
-            LogError("message handle is NULL");
-            result = IOTHUB_CLIENT_ERROR;
+            result = IOTHUB_CLIENT_OK;
```

Comment:
**Comment 1 - important** - How does IoT Hub handle it when app does a IOTHUBMESSAGE_ABANDONED|IOTHUBMESSAGE_REJECTED?  If device never responds is it going to be in some sort of infinite loop of trying to resend the message?

Same question with IoT Edge, since it's "hub/MQTT" implementation is slightly different than Hub's so this may not be same answer.

I get MQTT doesn't have a way to signal this but I worry a lot about infinite loop case, unless hub/Edge is fine with no ACK?

**COMMENT2 - nit** Assuming we're OK just eating this message, I feel that a comment (no need to LogInfo()) that MQTT doesn't define a way to NACK really so all we can do is to just not send data.  A non-MQTT expert - which SDK tries to support by hiding all this stuff - who had to debug into code for whatever reason may not appreciate that nuance at 1st.

Reply:
Comment 1: As per design, an ABANDONED message is requeued to be delivered to the device again when it reconnects. A REJECTED message is simply dequeued by the hub and never re-delivered.
If the customer does ABANDON a message, it is expected (no matter if using IoT Hub or Edge) that the message will be redelivered upon reconnection. So it will be up to the customer to control that.

Note: Currently for MQTT we do not send a PUBACK for a rejected message. I now believe that we should send  a PUBACK, to simulate the scenario of rejecting the message (so it does not get redelivered later).

Comment 2: Sure, that makes sense. Suggested comment:

```c
// Azure IoT Hub (and Edge) support the following responses for devicebound messages:
// ACCEPTED: message shall be dequeued by the hub and consider the delivery completed.
// ABANDONED: message shall remain in the hub queue, and re-delivered upon client re-connection.
// REJECTED: message shall be dequeued by the hub and not not re-delivered anymore.
// The MQTT protocol by design only supports acknowledging the receipt of a message (through PUBACK).
// To simulate the message responses above, this function behaves as follows:
// ACCEPTED: a PUBACK is sent to the Hub/Edge.
// ABANDONED: no response is sent to the Hub/Edge.
// REJECTED: a PUBACK is sent to the Hub/Edge (same as for ACCEPTED).
```